File: push-be/internal/domain/pagination.go
package domain

import (
	"encoding/base64"
	"fmt"
	"strings"
	"time"

	"github.com/google/uuid"
)

const (
	DefaultLimit = 50
	MaxLimit     = 100
)

type Cursor struct {
	CreatedAt time.Time
	ID        uuid.UUID
}

func (c Cursor) Encode() string {
	raw := c.CreatedAt.UTC().Format(time.RFC3339Nano) + "|" + c.ID.String()
	return base64.RawURLEncoding.EncodeToString([]byte(raw))
}

func DecodeCursor(s string) (Cursor, error) {
	raw, err := base64.RawURLEncoding.DecodeString(s)
	if err != nil {
		return Cursor{}, err
	}
	parts := strings.SplitN(string(raw), "|", 2)
	if len(parts) != 2 {
		return Cursor{}, fmt.Errorf("bad cursor")
	}
	ts, err := time.Parse(time.RFC3339Nano, parts[0])
	if err != nil {
		return Cursor{}, err
	}
	id, err := uuid.Parse(parts[1])
	if err != nil {
		return Cursor{}, err
	}
	return Cursor{CreatedAt: ts, ID: id}, nil
}

type PageRequest struct {
	Limit  int
	Cursor *Cursor
}

func (p PageRequest) EffectiveLimit() int {
	if p.Limit <= 0 {
		return DefaultLimit
	}
	if p.Limit > MaxLimit {
		return MaxLimit
	}
	return p.Limit
}

type Page[T any] struct {
	Items      []T
	NextCursor *string
	HasMore    bool
}

func NewPage[T any](items []T, limit int, cursorOf func(T) Cursor) Page[T] {
	p := Page[T]{Items: items}
	if len(items) > limit {
		p.Items = items[:limit]
		p.HasMore = true
		c := cursorOf(p.Items[len(p.Items)-1]).Encode()
		p.NextCursor = &c
	}
	if p.Items == nil {
		p.Items = []T{}
	}
	return p
}

File: push-be/app/domain/pagination.py
from __future__ import annotations
import base64
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Callable, Generic, Optional, TypeVar
from uuid import UUID

DEFAULT_LIMIT = 50
MAX_LIMIT = 100

T = TypeVar("T")


@dataclass
class Cursor:
    created_at: datetime
    id: UUID

    def encode(self) -> str:
        ts = self.created_at.astimezone(timezone.utc)
        frac = ""
        if ts.microsecond:
            frac = f".{ts.microsecond:06d}".rstrip("0")
        raw = ts.strftime("%Y-%m-%dT%H:%M:%S") + frac + "Z|" + str(self.id)
        return base64.urlsafe_b64encode(raw.encode()).rstrip(b"=").decode()


def decode_cursor(s: str) -> Cursor:
    try:
        raw = base64.urlsafe_b64decode(s + "=" * (-len(s) % 4))
        parts = raw.decode().split("|", 1)
        if len(parts) != 2:
            raise ValueError("bad cursor")
        return Cursor(created_at=datetime.fromisoformat(parts[0]), id=UUID(parts[1]))
    except Exception:
        raise ValueError("bad cursor")


@dataclass
class PageRequest:
    limit: int = 0
    cursor: Optional[Cursor] = None

    def effective_limit(self) -> int:
        if self.limit <= 0:
            return DEFAULT_LIMIT
        return min(self.limit, MAX_LIMIT)


@dataclass
class Page(Generic[T]):
    items: list[T] = field(default_factory=list)
    next_cursor: Optional[str] = None
    has_more: bool = False


def new_page(items: list[T], limit: int, cursor_of: Callable[[T], Cursor]) -> Page[T]:
    p: Page[T] = Page(items=items or [])
    if len(p.items) > limit:
        p.items = p.items[:limit]
        p.has_more = True
        p.next_cursor = cursor_of(p.items[-1]).encode()
    return p
